{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Mitovirus / Narnaviridae Set-up\n",
    "```\n",
    "Lead     : ababaian\n",
    "Issue    : NA\n",
    "start    : 2020 10 12\n",
    "complete : 2020 10 26\n",
    "files    : ~/serratus/notebook/201012_ab/\n",
    "s3 files : s3://serratus-public/notebook/YYMMDD/\n",
    "```\n",
    "\n",
    "## Introduction\n",
    "\n",
    "I recieved email from Adam Bergman and Samantha Lewis at UC Berkeley, in brief they are interested in a potential collaboration to apply Serratus for searching for novel Mitoviruses within Metazoa, specifically human. These are a class of viruses characterized as infecting the mitochondria of fungi.\n",
    "\n",
    "In brief, [Mitoviruses](https://viralzone.expasy.org/304) are short (2-3 kb) ssRNA+ viruses encoding for an RdRp. They fall within the family Narnaviridae.\n",
    "\n",
    "There are several routes with which this question can be explored from a technical standpoint. The greatest 'complication' arises from the difference in the vertrebrate genetic code and mitochondrial genetic code\n",
    "\n",
    "- [Mitovirus Taxa: 186768](https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?id=186768)\n",
    "- [Narnavirus Taxa: 186766](https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?mode=Tree&id=186766)\n",
    "- [Lenarviricota Phylum: 2732407](https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?mode=Info&id=2732407)\n",
    "\n",
    "**Standard Genetic Code (1)**\n",
    "```\n",
    "   AAs  = FFLLSSSSYY**CC*WLLLLPPPPHHQQRRRRIIIMTTTTNNKKSSRRVVVVAAAADDEEGGGG\n",
    "  Starts = ---M------**--*----M---------------M----------------------------\n",
    "  Base1  = TTTTTTTTTTTTTTTTCCCCCCCCCCCCCCCCAAAAAAAAAAAAAAAAGGGGGGGGGGGGGGGG\n",
    "  Base2  = TTTTCCCCAAAAGGGGTTTTCCCCAAAAGGGGTTTTCCCCAAAAGGGGTTTTCCCCAAAAGGGG\n",
    "  Base3  = TCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAG\n",
    "```\n",
    "\n",
    "**Vertebrate Mitochondria (2)**\n",
    "```\n",
    "        Code 2          Standard\n",
    "\n",
    " AGA    Ter  *          Arg  R\n",
    " AGG    Ter  *          Arg  R\n",
    " AUA    Met  M          Ile  I\n",
    " UGA    Trp  W          Ter  *\n",
    "```\n",
    "\n",
    "**Fungi Mitrochondria (4)**\n",
    "```\n",
    "      Code 4         Standard\n",
    "\n",
    " UGA    Trp  W          Ter  *\n",
    "```\n",
    "\n",
    "**Genetic Code**\n",
    "![Genetic code](201012_ab/genetic_code.jpg)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Spearfish Method\n",
    "\n",
    "The direct method will be to use the RdRP protein sequences from the known Mitovirus (gc 4) and Narnavirus (gc standard) outgroup with their respective genetic codes. These protein sequence can then be searched with `diamond` using a code 2 (vertebrate MT) translation.\n",
    "\n",
    "This is the most direct and specific means with which to search for a Mito/Narna protein sequence in vertebrate samples. This search is not compatible with any other sequences (i.e. protref5) since the altered translation  in diamond will render other standard RdRp translations incorrect.\n",
    "\n",
    "The advantage of including other (non-narna) sequences though is if there are mitochondrially encoded RdRp in other families they may result in a hit.\n",
    "\n",
    "### Trolling Method\n",
    "\n",
    "An indirect method is to extract the genetic code appropriate CDS sequence from each Mitovirus/Narnavirus and force translation via the standard code. Assuming a hypothetical vertebrate Mitovirus would use mtDNA code,the following changes would apply:\n",
    "\n",
    "```\n",
    "        Code 2          Standard\n",
    "\n",
    " AGA    Ter  *          Arg  R\n",
    " AGG    Ter  *          Arg  R\n",
    " AUA    Met  M          Ile  I\n",
    " UGA    Trp  W          Ter  *\n",
    "```\n",
    "\n",
    "The TER --> ARG changes are perfectly acceptable, these will be defined by CDS boundaries.\n",
    "\n",
    "The AUA M --> I substitution is a `+1` in `BLOSUM62` instead of M match `+5` so it won't significantly alter the score (not a loss). \n",
    "\n",
    "The primary alteration is UGA (W) will be interpreted as a stop codon by the standard code, which I can manually change via grep to \\*.\n",
    "\n",
    "There is an option  `-l1` which should be capable of using these translated sequences, from `diamond` docs:\n",
    "\n",
    "\n",
    ">`--min-orf/-l #`\n",
    ">\n",
    ">Ignore translated sequences that do not contain an open reading frame of at least this length. By default this feature is disabled for sequences of length below 30, set to 20 for sequences of length below 100, and set to 40 otherwise. Setting this option to 1 will disable this feature.\n",
    "\n",
    "Overall this design may lead to some false-positives where mt-encoded stop codons are read as \"R\" by diamond and some false-negatives where the stop-codon "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Materials and Methods\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reference Sequences (Nucleotide)\n",
    "\n",
    "#### mitovirus.fa\n",
    "\n",
    "Query: `txid186768[Organism:exp]`\n",
    "Date : `2020-10-12`\n",
    "Results: `2364`\n",
    "\n",
    "Downloaded complete sequences, `mitovirus.fa` as well as\"nucleotide CDS\", `mitovirus.cds.fa`\n",
    "\n",
    "\n",
    "#### lenarviricota.fa (call it narnavirus for simplicity)\n",
    "\n",
    "Query: `txid2732407[Organism:exp] NOT txid186768[Organism:exp]`\n",
    "Date : `2020-10-12`\n",
    "Results: `4878`\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sun Oct 25 15:51:15 PDT 2020\n"
     ]
    }
   ],
   "source": [
    "# Serratus commit version\n",
    "SERRATUS=\"/home/artem/serratus\"\n",
    "cd $SERRATUS\n",
    "\n",
    "# Create local run directory\n",
    "WORK=\"$SERRATUS/notebook/201012_ab\"\n",
    "mkdir -p $WORK; cd $WORK\n",
    "\n",
    "# S3 notebook path\n",
    "S3_WORK='s3://serratus-public/notebook/201012_ab/'\n",
    "\n",
    "# date and version\n",
    "date"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[INFO]\u001b[0m read sequences ...\n",
      "[INFO]\u001b[0m 2200 sequences loaded\n",
      "[INFO]\u001b[0m sorting ...\n",
      "[INFO]\u001b[0m output ...\n",
      "[INFO]\u001b[0m 45 duplicated records removed\n"
     ]
    }
   ],
   "source": [
    "# Mitovirus\n",
    "# EMBOSS 6.6.0\n",
    "# Retrieve CDS sequence from mitovirus from nuccore\n",
    "\n",
    "NAME='mitovirus'\n",
    "seqkit sort -l -r mitovirus.cds.fa > mitovirus.cds.1.fa\n",
    "\n",
    "# Remove hypothetical protein cds\n",
    "seqkit grep -n -v -r -p \"hypo\" mitovirus.cds.1.fa \\\n",
    "  > mitovirus.cds.2.fa\n",
    "\n",
    "# Save full headers\n",
    "grep \">\" mitovirus.cds.2.fa > mitovirus.cds.headers\n",
    "\n",
    "# Reduce header to accession name\n",
    "sed 's/lcl|//g' mitovirus.cds.2.fa \\\n",
    "  | sed 's/_cds.*//g' \\\n",
    "  > mitovirus.cds.3.fa\n",
    "  \n",
    "# Remove duplicates (in effect, keep only longest ORF from each file)\n",
    "seqkit rmdup -n mitovirus.cds.3.fa >  mitovirus.rdrp.fa\n",
    "\n",
    "samtools faidx mitrovirus.rdrp.fa\n",
    "\n",
    "# Final: Mitovirus.RdRP.MN036117.1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# EMBOSS 6.6.0\n",
    "# Translate sequence using standard genetic code\n",
    "NAME='mitovirus'\n",
    "\n",
    "transeq \\\n",
    "  -sequence mitovirus.rdrp.fa \\\n",
    "  -frame 1 \\\n",
    "  -table 0 \\\n",
    "  -methionine \\\n",
    "  -outseq mitovirus.rdrp.aa.fa\n",
    "  \n",
    "sed -i 's/._.//g' mitovirus.rdrp.aa.fa\n",
    "sed -i 's/>/>Mitovirus.rdrp./g' mitovirus.rdrp.aa.fa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Lenarviricota.cds.fa\n",
    "# EMBOSS 6.6.0\n",
    "# Retrieve CDS sequence from mitovirus from nuccore\n",
    "cd $WORK\n",
    "\n",
    "NAME='lenarviricota'\n",
    "seqkit sort -l -r $NAME.cds.fa > $NAME.cds.1.fa\n",
    "\n",
    "# Remove hypothetical protein cds\n",
    "seqkit grep -n -v -r -p \"hypo\" $NAME.cds.1.fa \\\n",
    "  > $NAME.cds.2.fa\n",
    "\n",
    "# Save full headers\n",
    "grep \">\" $NAME.cds.2.fa > $NAME.cds.headers\n",
    "\n",
    "# Reduce header to accession name\n",
    "sed 's/lcl|//g' $NAME.cds.2.fa \\\n",
    "  | sed 's/_cds.*//g' \\\n",
    "  > $NAME.cds.3.fa\n",
    "  \n",
    "# Remove duplicates (in effect, keep only longest ORF from each file)\n",
    "seqkit rmdup -n $NAME.cds.3.fa >  $NAME.rdrp.fa\n",
    "\n",
    "samtools faidx $NAME.rdrp.fa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "transeq \\\n",
    "  -sequence $NAME.rdrp.fa \\\n",
    "  -frame 1 \\\n",
    "  -table 0 \\\n",
    "  -methionine \\\n",
    "  -outseq $NAME.rdrp.aa.fa\n",
    "  \n",
    "sed -i 's/._.//g' $NAME.rdrp.aa.fa\n",
    "sed -i 's/>/>Narnavirus.rdrp./g' $NAME.rdrp.aa.fa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd $WORK\n",
    "\n",
    "# Clean-up folders\n",
    "mkdir mito\n",
    "mv mitov* mito/\n",
    "\n",
    "mkdir lena\n",
    "mv lenar* lena/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usearch v11.0.667_i86linux32, 4.0Gb RAM (16.3Gb total), 4 cores\n",
      "(C) Copyright 2013-18 Robert C. Edgar, all rights reserved.\n",
      "https://drive5.com/usearch\n",
      "\n",
      "License: personal use only\n",
      "\n",
      "00:02 132Mb   100.0% 1948 clusters, max size 444, avg 1.8\n",
      "00:02 132Mb   100.0% Writing centroids to narnavirus.id95.fa\n",
      "                                                            \n",
      "      Seqs  3422\n",
      "  Clusters  1948\n",
      "  Max size  444\n",
      "  Avg size  1.8\n",
      "  Min size  1\n",
      "Singletons  1620, 47.3% of seqs, 83.2% of clusters\n",
      "   Max mem  132Mb\n",
      "      Time  2.00s\n",
      "Throughput  1711.0 seqs/sec.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Create protein refernece at 95%\n",
    "NAME='narnavirus'\n",
    "\n",
    "# Sort priority\n",
    "# Merge and cluster for protref add-on\n",
    "cat mito/mitovirus.rdrp.aa.fa lena/lenarviricota.rdrp.aa.fa \\\n",
    "  > $NAME.aa.fa\n",
    "\n",
    "# Prune to 95% nucleotide identity\n",
    "usearch -cluster_smallmem narna.aa.fa \\\n",
    "   -id 0.95 \\\n",
    "   -sortedby other \\\n",
    "   -maxaccepts 4 \\\n",
    "   -maxrejects 64 \\\n",
    "   -maxhits 1 \\\n",
    "   -uc $NAME.id95.uc \\\n",
    "   -centroids $NAME.id95.fa\n",
    "   \n",
    "grep \"^>\" $NAME.id95.fa > $NAME.id95.headers\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "upload: ./genetic_code.jpg to s3://serratus-public/notebook/201012_ab/genetic_code.jpg\n",
      "upload: lena/lenarviricota.rdrp.fa.fai to s3://serratus-public/notebook/201012_ab/lena/lenarviricota.rdrp.fa.fai\n",
      "upload: lena/lenarviricota.cds.headers to s3://serratus-public/notebook/201012_ab/lena/lenarviricota.cds.headers\n",
      "upload: mito/mitovirus.cds.headers to s3://serratus-public/notebook/201012_ab/mito/mitovirus.cds.headers\n",
      "upload: lena/lenarviricota.rdrp.aa.fa to s3://serratus-public/notebook/201012_ab/lena/lenarviricota.rdrp.aa.fa\n",
      "upload: mito/mitovirus.cds.1.fa to s3://serratus-public/notebook/201012_ab/mito/mitovirus.cds.1.fa\n",
      "upload: mito/mitovirus.cds.2.fa to s3://serratus-public/notebook/201012_ab/mito/mitovirus.cds.2.fa\n",
      "upload: mito/mitovirus.rdrp.aa.fa to s3://serratus-public/notebook/201012_ab/mito/mitovirus.rdrp.aa.fa\n",
      "upload: mito/mitovirus.cds.3.fa to s3://serratus-public/notebook/201012_ab/mito/mitovirus.cds.3.fa\n",
      "upload: mito/mitovirus.rdrp.fa.fai to s3://serratus-public/notebook/201012_ab/mito/mitovirus.rdrp.fa.fai\n",
      "upload: ./mtDNA_code.jpg to s3://serratus-public/notebook/201012_ab/mtDNA_code.jpg\n",
      "upload: lena/lenarviricota.rdrp.fa to s3://serratus-public/notebook/201012_ab/lena/lenarviricota.rdrp.fa\n",
      "upload: mito/mitovirus.cds.fa to s3://serratus-public/notebook/201012_ab/mito/mitovirus.cds.fa\n",
      "upload: lena/lenarviricota.cds.3.fa to s3://serratus-public/notebook/201012_ab/lena/lenarviricota.cds.3.fa\n",
      "upload: ./narnavirus.id95.uc to s3://serratus-public/notebook/201012_ab/narnavirus.id95.uc\n",
      "upload: ./narnavirus.id95.headers to s3://serratus-public/notebook/201012_ab/narnavirus.id95.headers\n",
      "upload: mito/mitovirus.rdrp.fa to s3://serratus-public/notebook/201012_ab/mito/mitovirus.rdrp.fa\n",
      "upload: ./narnavirus.aa.fa to s3://serratus-public/notebook/201012_ab/narnavirus.aa.fa\n",
      "upload: mito/mitovirus.fa to s3://serratus-public/notebook/201012_ab/mito/mitovirus.fa\n",
      "upload: ./narnavirus.id95.fa to s3://serratus-public/notebook/201012_ab/narnavirus.id95.fa\n",
      "upload: ./narna.aa.fa to s3://serratus-public/notebook/201012_ab/narna.aa.fa\n",
      "upload: lena/lenarviricota.cds.2.fa to s3://serratus-public/notebook/201012_ab/lena/lenarviricota.cds.2.fa\n",
      "upload: lena/lenarviricota.fa to s3://serratus-public/notebook/201012_ab/lena/lenarviricota.fa\n",
      "upload: lena/lenarviricota.cds.1.fa to s3://serratus-public/notebook/201012_ab/lena/lenarviricota.cds.1.fa\n",
      "upload: lena/lenarviricota.cds.fa to s3://serratus-public/notebook/201012_ab/lena/lenarviricota.cds.fa\n"
     ]
    }
   ],
   "source": [
    "aws s3 sync ./ $S3_WORK"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fire up EC2 Instance\n",
    "sudo yum install -y docker\n",
    "sudo yum install -y git\n",
    "sudo service docker start\n",
    "\n",
    "# Download latest serratus repo\n",
    "git clone -b diamond-dev https://github.com/ababaian/serratus.git; cd serratus/containers\n",
    "\n",
    "# If you want to upload containers to your repository, include this.\n",
    "export DOCKERHUB_USER='serratusbio' # optional\n",
    "sudo docker login # optional\n",
    "\n",
    "# Build all containers and upload them docker hub repo (if available)\n",
    "./build_containers.sh\n",
    "\n",
    "# Launch aligner container\n",
    "sudo docker run --rm --entrypoint /bin/bash -it serratus-align:latest\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Nucleotide Sequence (CDS)\n",
    "```\n",
    ">HE586976.1\n",
    "ATTGTAGCTATGTTAGATTATACAACACAGCTATTTCTTCGACCTATACATTCTGACTTG\n",
    "TTTAAACTTCTAAAAAAGTTACCACAAGATAGAACTTTTACCCAAAATCCATTAAATGAT\n",
    "TGAGAAGATAATGAACACTCATTTTGATCAATCGACCTTACAGCTGCAACTGATAGATTT\n",
    "CCCATTAGTTTACAACGGCGGTTATTACTATATATATATAGTGATCCCGAAATTGCAAAT\n",
    "TCTTGGCAAAATCTATTAGTACATAGAGAATATGCCCGTAATGGGTTAAGTCCAATAAAA\n",
    "TATTCTGTTGGACAGCCCATGGGAGCATATTCATCCTGACCTGCTTTCACATTATCTCAT\n",
    "CACCTTGTAGTTCATTGATGTGCACATTTATGCAACATCAATAAATTCAAGGATTATATA\n",
    "ATTCTTGGTGACGATATTGTTATACATAACGATAACGTTGCTAAAAAATATATTGAAATA\n",
    "ATGGGAAAATTAGGAGTGGGTCTATCAGATAGTAAAACACATGTATCAAAAG\n",
    "```\n",
    "Protein Sequence\n",
    "```\n",
    ">Mitovirus.rdrp.HE586976\n",
    "IVAMLDYTTQLFLRPIHSDLFKLLKKLPQDRTFTQNPLND*EDNEHSF*SIDLTAATDRF\n",
    "PISLQRRLLLYIYSDPEIANSWQNLLVHREYARNGLSPIKYSVGQPMGAYSS*PAFTLSH\n",
    "HLVVH*CAHLCNINKFKDYIILGDDIVIHNDNVAKKYIEIMGKLGVGLSDSKTHVSKX\n",
    "```\n",
    "\n",
    "Pseudo-reads\n",
    "```\n",
    ">r1\n",
    "ATTGTAGCTATGTTAGATTATACAACACAGCTATTTCTTCGACCTATACATTCTGACTTG\n",
    ">r2\n",
    "TTTAAACTTCTAAAAAAGTTACCACAAGATAGAACTTTTACCCAAAATCCATTAAATGAT\n",
    ">r3\n",
    "TGAGAAGATAATGAACACTCATTTTGATCAATCGACCTTACAGCTGCAACTGATAGATTT\n",
    ">r3_firststop\n",
    "TGTGAAGATAATGAACACTCATTTTGATCAATCGACCTTACAGCTGCAACTGATAGATTT\n",
    ">r5\n",
    "CCCATTAGTTTACAACGGCGGTTATTACTATATATATATAGTGATCCCGAAATTGCAAAT\n",
    ">r6\n",
    "TCTTGGCAAAATCTATTAGTACATAGAGAATATGCCCGTAATGGGTTAAGTCCAATAAAA\n",
    ">r7\n",
    "TATTCTGTTGGACAGCCCATGGGAGCATATTCATCCTGACCTGCTTTCACATTATCTCAT\n",
    ">r8\n",
    "CACCTTGTAGTTCATTGATGTGCACATTTATGCAACATCAATAAATTCAAGGATTATATA\n",
    ">r9\n",
    "ATTCTTGGTGACGATATTGTTATACATAACGATAACGTTGCTAAAAAATATATTGAAATA\n",
    "```\n",
    "\n",
    "Diamond Test\n",
    "```\n",
    "r1      Mitovirus.rdrp.HE586971 1       60      60      1       20      178     100.0   7.0e-07 20      20M     +       ATTGTAGCTATGTTAGATTATACAACACAGCTATTTCTTCGACCTATACATTCTGACTTG    IVAMLDYTTQLFLRPIHSDL\n",
    "r2      Mitovirus.rdrp.AF534641 1       60      60      266     285     742     100.0   3.1e-07 20      20M     +       TTTAAACTTCTAAAAAAGTTACCACAAGATAGAACTTTTACCCAAAATCCATTAAATGAT    FKLLKKLPQDRTFTQNPLND\n",
    "r3      Mitovirus.rdrp.AF534641 1       60      60      286     305     742     100.0   1.6e-06 20      20M     +       TGAGAAGATAATGAACACTCATTTTGATCAATCGACCTTACAGCTGCAACTGATAGATTT    *EDNEHSF*SIDLTAATDRF\n",
    "r5      Mitovirus.rdrp.AF534641 1       60      60      306     325     742     100.0   9.1e-07 20      20M     +       CCCATTAGTTTACAACGGCGGTTATTACTATATATATATAGTGATCCCGAAATTGCAAAT    PISLQRRLLLYIYSDPEIAN\n",
    "r6      Mitovirus.rdrp.HE586971 1       60      60      81      100     178     100.0   6.3e-08 20      20M     +       TCTTGGCAAAATCTATTAGTACATAGAGAATATGCCCGTAATGGGTTAAGTCCAATAAAA    SWQNLLVHREYARNGLSPIK\n",
    "r7      Mitovirus.rdrp.AF534641 1       60      60      346     365     742     100.0   4.1e-07 20      20M     +       TATTCTGTTGGACAGCCCATGGGAGCATATTCATCCTGACCTGCTTTCACATTATCTCAT    YSVGQPMGAYSS*PAFTLSH\n",
    "r8      Mitovirus.rdrp.AF534641 1       60      60      366     385     742     100.0   2.2e-08 20      20M     +       CACCTTGTAGTTCATTGATGTGCACATTTATGCAACATCAATAAATTCAAGGATTATATA    HLVVH*CAHLCNINKFKDYI\n",
    "r9      Mitovirus.rdrp.HE586971 1       60      60      141     160     178     100.0   1.2e-06 20      20M     +       ATTCTTGGTGACGATATTGTTATACATAACGATAACGTTGCTAAAAAATATATTGAAATA    ILGDDIVIHNDNVAKKYIEI\n",
    "```\n",
    "\n",
    "From this it looks like this set-up will work out of the box with diamond, it's handling stop codons quite efficiently.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# In serratus-align\n",
    "mkdir tmp; cd tmp\n",
    "aws s3 cp s3://serratus-public/notebook/201012_ab/mitovirus.rdrp.aa.fa ./mito.fa\n",
    "\n",
    "# Make mitovirus database\n",
    "diamond makedb --in mito.fa -d mito\n",
    "\n",
    "# Test diamond search with interrupting stop codons\n",
    "# based on standard genetic code\n",
    "\n",
    "cat reads.fa \\\n",
    "  | diamond blastx \\\n",
    "   -d mito.dmnd \\\n",
    "   --unal 0 \\\n",
    "   -k 1 \\\n",
    "   -p 1 \\\n",
    "   -b 0.2 \\\n",
    "   -f 6 qseqid sseqid qstart qend qlen sstart send slen pident evalue btop cigar qstrand qseq sseq \\\n",
    "   > tmp.bam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Update to protref5b"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mkdir p5b; cd p5b\n",
    "aws s3 sync s3://serratus-public/seq/protref5/ ./\n",
    "aws s3 cp $S3_WORK/narnavirus.id95.fa ./\n",
    "\n",
    "mkdir update; cd update\n",
    "cat ../protref5.fa ../narnavirus.id95.fa > protref5b.fa\n",
    "cp ../protref5.msa protref5b.msa\n",
    "\n",
    "# Make fasta index\n",
    "samtools faidx protref5b.fa\n",
    "mv protref5b.fa.fai protref5b.sumzer.tsv\n",
    "\n",
    "# Make diamond index\n",
    "diamond makedb --in protref5b.fa -d protref5b\n",
    "\n",
    "md5sum * > protref5b.md5\n",
    "\n",
    "# UPLOAD\n",
    " aws s3 sync ./ s3://serratus-public/seq/protref5b/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "3426e5b4d178776b0b5737417372bf5a  protref5b.dmnd\n",
    "55b7ac3cd282703e4a512b36361f7da5  protref5b.fa\n",
    "0da30de8910dd6bdb3c7f150dc06f65b  protref5b.md5\n",
    "e094fc7db19c07ffcedf8bc42963ab80  protref5b.msa\n",
    "96c81da14ae47216ec9267ae0e2d1b42  protref5b.sumzer.tsv\n",
    "```"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
