{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# SRA Run Info Update \n",
    "```\n",
    "Lead     : ababaian\n",
    "Issue    : <github issue #>\n",
    "start    : 2020 12 30\n",
    "complete : 2021 01 02\n",
    "files    : ~/serratus/notebook/201230_sra/\n",
    "s3 files : s3://serratus-public/notebook/201230_sra/\n",
    "```\n",
    "\n",
    "## Introduction\n",
    "\n",
    "It has been about half a year since the last big batches of SRA Runs. This will be an update of the main \"SRA\" files for more runs in the future. (Exponential growth and all!)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Initialization\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wed Dec 30 09:01:24 PST 2020\n",
      "05de2c7adfd6951e98261ff5701e9303aed15cbc\n"
     ]
    }
   ],
   "source": [
    "# Serratus commit version\n",
    "SERRATUS=\"/home/artem/serratus\"\n",
    "cd $SERRATUS\n",
    "\n",
    "# Create local run directory\n",
    "WORK=\"$SERRATUS/notebook/201230_sra\"\n",
    "mkdir -p $WORK; cd $WORK\n",
    "\n",
    "# S3 notebook path\n",
    "S3_WORK='s3://serratus-public/notebook/201230_sra'\n",
    "\n",
    "# date and version\n",
    "date\n",
    "git rev-parse HEAD # commit version"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Updates\n",
    "\n",
    "(See previous RunInfo files below)\n",
    "\n",
    "Files downloaded on wired Raspberry Pi connection at home."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Total Unique: ` 5 780 800`\n",
    "\n",
    "### Human (non-controlled data)\n",
    "\n",
    "- File: `hu_SraRunInfo.csv`\n",
    "\n",
    "Search Term:\n",
    "```\n",
    "\"Homo sapiens\"[Organism] AND (\"type_rnaseq\"[Filter]) AND cluster_public[prop] AND \"platform illumina\"[Properties]\n",
    "```\n",
    "SRA Accessed: `2020/12/30`\n",
    "\n",
    "Results: `789 931`\n",
    "\n",
    "**AND**\n",
    "\n",
    "- File: `hu_meta_SraRunInfo.csv`\n",
    "\n",
    "Search Term: \n",
    "```\n",
    "\"Homo sapiens\"[Organism] AND (\"metagenomic\"[Filter] OR \"metatranscriptomic\"[Filter]) AND cluster_public[prop] AND \"platform illumina\"[Properties]\n",
    "```\n",
    "SRA Accessed: `2020/12/30`\n",
    "\n",
    "Results: `47 763`\n",
    "\n",
    "### Mouse\n",
    "\n",
    "- File: `mu_SraRunInfo.csv`\n",
    "\n",
    "Search Term:\n",
    "```\n",
    "\"Mus musculus\"[Organism] AND (\"type_rnaseq\"[Filter] OR \"metagenomic\"[Filter] OR \"metatranscriptomic\"[Filter]) AND \"platform illumina\"[Properties]\n",
    "```\n",
    "\n",
    "SRA Accessed: `2020/12/30` \n",
    "\n",
    "Results: `1 058 559`\n",
    "\n",
    "### Mammalian\n",
    "\n",
    "- File: `mamm_SraRunInfo.csv`\n",
    "\n",
    "Search Term:\n",
    "```\n",
    "(\"Mammalia\"[Organism] NOT \"Homo sapiens\"[Organism] NOT \"Mus musculus\"[orgn]) AND (\"type_rnaseq\"[Filter] OR \"metagenomic\"[Filter] OR \"metatranscriptomic\"[Filter]) AND \"platform illumina\"[Properties]\n",
    "```\n",
    "SRA Accessed: `2020/12/30`\n",
    "\n",
    "Results: `126 382`\n",
    "\n",
    "### Vertebrate\n",
    "\n",
    "- File: `vert_SraRunInfo.csv`\n",
    "\n",
    "Search Term: \n",
    "```\n",
    "(\"Vertebrata\"[Organism] NOT \"Mammalia\"[Organism] NOT \"Homo sapiens\"[Organism] NOT \"Mus musculus\"[orgn]) AND (\"type_rnaseq\"[Filter] OR \"metagenomic\"[Filter] OR \"metatranscriptomic\"[Filter]) AND cluster_public[prop] AND \"platform illumina\"[Properties]\n",
    "```\n",
    "SRA Accessed: `2020/12/30`\n",
    "\n",
    "Results: `114 078`\n",
    "\n",
    "### Invertebrate\n",
    "\n",
    "- File: `inv_SraRunInfo.csv`\n",
    "\n",
    "Search Term: \n",
    "```\n",
    "(\"Metazoa\"[Organism] NOT \"Vertebrata\"[Organism]) (\"type_rnaseq\"[Filter] OR \"metagenomic\"[Filter] OR \"metatranscriptomic\"[Filter]) AND cluster_public[prop] AND \"platform illumina\"[Properties]\n",
    "```\n",
    "\n",
    "SRA Accessed: `2020/12/31`\n",
    "\n",
    "Results: `184 729`\n",
    "\n",
    "\n",
    "### Eukarya\n",
    "\n",
    "- File: `euk_SraRunInfo.csv`\n",
    "\n",
    "Search Term: \n",
    "```\n",
    "(\"Eukaryota\"[Organism] NOT \"Metazoa\"[Organism]) AND (\"type_rnaseq\"[Filter] OR \"metagenomic\"[Filter] OR \"metatranscriptomic\"[Filter]) AND cluster_public[prop] AND \"platform illumina\"[Properties]\n",
    "```\n",
    "\n",
    "SRA Accessed: `2021/01/01`\n",
    "\n",
    "Results: `367 924`\n",
    "\n",
    "\n",
    "### Prokaya / Other\n",
    "\n",
    "- File: `pro_SraRunInfo.csv`\n",
    "\n",
    "Search Term: \n",
    "```\n",
    "(\"type_rnaseq\"[Filter] OR \"RNASEQ\" OR \"metagenomic\"[Filter] OR \"METAGENOME\" OR \"metatranscriptomic\"[Filter] OR \"METATRANSCRIPTOME\" OR \"ENVIRONMENTAL\") NOT \"Eukaryota\"[Organism] AND cluster_public[prop] AND \"platform illumina\"[Properties]\n",
    "```\n",
    "\n",
    "SRA Accessed: `2021/01/20`\n",
    "\n",
    "Results: `2 672 802`\n",
    "\n",
    "### Virome\n",
    "\n",
    "- File: `viro_SraRunInfo.csv`\n",
    "\n",
    "Search Term: \n",
    "```\n",
    "\"VIRAL METAGENOME\" OR \"VIROME\" OR \"VIROMIC\" OR \"VIRAL RNA\" OR \"VIRAL METAGENOMICS\" NOT amplicon[All Fields] AND \"platform illumina\"[Properties] AND cluster_public[prop]\n",
    "```\n",
    "\n",
    "SRA Accessed: `2020/12/31`\n",
    "\n",
    "Results: `52 072`\n",
    "\n",
    "\n",
    "### Metagenomic\n",
    "\n",
    "- File: `meta_SraRunInfo.csv`\n",
    "\n",
    "Search Term: \n",
    "```\n",
    "\"METAGENOMIC\" NOT amplicon[All Fields] AND \"platform illumina\"[Properties] AND cluster_public[prop]\n",
    "```\n",
    "\n",
    "SRA Accessed: `2020/12/31`\n",
    "\n",
    "Results: `566 826`\n",
    "\n",
    "### Bat (with WGS)\n",
    "\n",
    "- File: `bat_SraRunInfo.csv`\n",
    "\n",
    "Search Term: \n",
    "```\n",
    "\"bat\" OR txid9397[Organism:exp] AND \"platform illumina\"[Properties]\n",
    "```\n",
    "\n",
    "SRA Accessed: `2021/01/01`\n",
    "\n",
    "Results: `14 103`\n",
    "\n",
    "### WGS (mammalian, non-mouse)\n",
    "\n",
    "- File: `wgs_SraRunInfo.csv`\n",
    "\n",
    "Search Term:\n",
    "```\n",
    "(\"Mammalia\"[Organism] NOT \"Mus musculus\"[orgn]) AND (\"type_exome\"[Filter] OR \"type_genome\"[Filter]) NOT \"METAGENOME\" AND \"platform illumina\"[Properties] AND cluster_public[prop]\n",
    "```\n",
    "SRA Accessed: `2021/01/21`\n",
    "\n",
    "Results: `393 278`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     14103 bat_SraRunInfo.csv\n",
      "    367924 euk_SraRunInfo.csv\n",
      "    789931 hu_SraRunInfo.csv\n",
      "     47763 hu_meta_SraRunInfo.csv\n",
      "    184729 inv_SraRunInfo.csv\n",
      "    126382 mamm_SraRunInfo.csv\n",
      "    566826 meta_SraRunInfo.csv\n",
      "   1058559 mu_SraRunInfo.csv\n",
      "   2672802 pro_SraRunInfo.csv\n",
      "    114078 vert_SraRunInfo.csv\n",
      "     52072 viro_SraRunInfo.csv\n",
      "    393278 wgs_SraRunInfo.csv\n",
      "   6388447 total\n",
      "41006f8fa8fbe3ce3861a726847c888f  bat_SraRunInfo.csv\n",
      "b1ead72c647be78e471f29428d38f252  euk_SraRunInfo.csv\n",
      "0e3424b9cbcb2249314dc1ce42fecf0c  hu_SraRunInfo.csv\n",
      "ff4f6839d2434c5ecbb0bdc4cbbb990c  hu_meta_SraRunInfo.csv\n",
      "0cdc0f966e6fbdd1a35a63b3260806de  inv_SraRunInfo.csv\n",
      "e0ada71c13d1d80776383d7c0120f58c  mamm_SraRunInfo.csv\n",
      "7ffe7e78bc7df55bca73eaac20a026e2  meta_SraRunInfo.csv\n",
      "245884d3ded9056bd278f357347b53bf  mu_SraRunInfo.csv\n",
      "47cfa68db178cd51d28a604e8ee83c14  pro_SraRunInfo.csv\n",
      "cf36fb9216bc478da656f00b193569aa  vert_SraRunInfo.csv\n",
      "21984f68344cee144268f9fc9ac703e8  viro_SraRunInfo.csv\n",
      "47313cfe06b2fce2ada17a894a72a278  wgs_SraRunInfo.csv\n"
     ]
    }
   ],
   "source": [
    "# wc / md5 / upload\n",
    "cd /home/artem/Desktop/serratus/notebook/201230_sra/\n",
    "\n",
    "wc -l *csv\n",
    "md5sum *csv > SraRunInfo.md5\n",
    "md5sum *csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "upload: ./SraRunInfo.md5 to s3://serratus-public/notebook/201230_sra/SraRunInfo.md5\n",
      "upload: ./completed.sra to s3://serratus-public/notebook/201230_sra/completed.sra\n",
      "upload: ./uniq.sra to s3://serratus-public/notebook/201230_sra/uniq.sra\n",
      "upload: complete/rpro.done to s3://serratus-public/notebook/201230_sra/complete/rpro.done\n",
      "upload: ./wgs_SraRunInfo.csv to s3://serratus-public/notebook/201230_sra/wgs_SraRunInfo.csv\n"
     ]
    }
   ],
   "source": [
    "aws s3 sync ./ s3://serratus-public/notebook/201230_sra/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## COMPLETED RUNS\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fri Jan 22 13:45:23 PST 2021\n",
      "2ddf24166fdded0bc89a96b2541bcbc7442f98f9\n"
     ]
    }
   ],
   "source": [
    "# Serratus commit version\n",
    "SERRATUS=\"/home/artem/serratus\"\n",
    "cd $SERRATUS\n",
    "\n",
    "# Create local run directory\n",
    "WORK=\"$SERRATUS/notebook/201230_sra\"\n",
    "mkdir -p $WORK; cd $WORK\n",
    "\n",
    "# S3 notebook path\n",
    "S3_WORK='s3://serratus-public/notebook/201230_sra'\n",
    "\n",
    "# date and version\n",
    "date\n",
    "git rev-parse HEAD # commit version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd $WORK\n",
    "mkdir -p complete; cd complete"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Download list of 210111_rVert\n",
    "aws s3 ls s3://serratus-public/out/210111_rVert/psummary/ \\\n",
    "  > rvert.done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Download list of 210112_rMamm\n",
    "aws s3 ls s3://serratus-public/out/210112_rMamm/psummary/ \\\n",
    "  > rmamm.done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Download list of 210113_rViro\n",
    "aws s3 ls s3://serratus-public/out/210113_rViro/psummary/ \\\n",
    "  > rviro.done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Download list of 210113_rInv\n",
    "aws s3 ls s3://serratus-public/out/210113_rInv/psummary/ \\\n",
    "  > rinv.done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Download list of 210114_rEuk\n",
    "aws s3 ls s3://serratus-public/out/210114_rEuk/psummary/ \\\n",
    "  > reuk.done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Download list of 210115_rMeta (and its typo, d'oh!)\n",
    "aws s3 ls s3://serratus-public/out/210115_rMeta/psummary/ \\\n",
    "  > rmeta.done\n",
    "aws s3 ls s3://serratus-public/out/210115_rEuk/psummary/ \\\n",
    "  >> rmeta.done  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Download list of 210116_rHu\n",
    "aws s3 ls s3://serratus-public/out/210116_rHu/psummary/ \\\n",
    "  > rhu.done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": []
    }
   ],
   "source": [
    "# Download list of 210116_rMu\n",
    "aws s3 ls s3://serratus-public/out/210116_rMu/psummary/ \\\n",
    "  > rmu.done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Download list of 210118_QC\n",
    "aws s3 ls s3://serratus-public/out/210118_rQC/psummary/ \\\n",
    "  > rqc.done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Download list of 210118_Pro\n",
    "aws s3 ls s3://serratus-public/out/210119_rPro/psummary/ \\\n",
    "  > rpro.done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Download list of 210118_Pro\n",
    "aws s3 ls s3://serratus-public/out/210121_rWgs/psummary/ \\\n",
    "  > rwgs.done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   365437 reuk.done\n",
      "   680939 rhu.done\n",
      "   184133 rinv.done\n",
      "   125386 rmamm.done\n",
      "   564278 rmeta.done\n",
      "  1051719 rmu.done\n",
      "  2126560 rpro.done\n",
      "   132165 rqc.done\n",
      "   113555 rvert.done\n",
      "    51765 rviro.done\n",
      "   330239 rwgs.done\n",
      "  5726176 total\n",
      "5686715 ../completed.sra\n"
     ]
    }
   ],
   "source": [
    "# 2021-01-11 23:22:37        126 DRR001820.psummary\n",
    "# >\n",
    "# DRR001820\n",
    "cd $WORK/complete/\n",
    "wc -l *.done\n",
    "\n",
    "cat *.done \\\n",
    "  | sed 's/.* //g' - \\\n",
    "  | sed 's/\\.psummary//g' - \\\n",
    "  | sed 's/^$//g' - \\\n",
    "  | sort -u \\\n",
    "  > ../completed.sra\n",
    "  \n",
    "wc -l ../completed.sra"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Killed\n"
     ]
    }
   ],
   "source": [
    "cd $WORK\n",
    "\n",
    "remove_done () {\n",
    "  COMPLETE=$1\n",
    "  SRA=$2\n",
    "  \n",
    "  # Sort unique SRA in the input RunInfo file\n",
    "  cut -f1 -d',' $SRA \\\n",
    "    | sort - \\\n",
    "    > inputsra.tmp\n",
    "  \n",
    "  # Ensure completed SRA list is sorted\n",
    "  sort $COMPLETE > complete.sort.tmp\n",
    "  \n",
    "  # Supress SRA-accessions only in file 1 or file 1+2\n",
    "  # Show SRA unique to file 2\n",
    "  # parse into a regex expression to find matching lines\n",
    "  comm -1 -3 complete.sort.tmp inputsra.tmp \\\n",
    "    | sed 's/^$\\n//g' - \\\n",
    "    | sed 's/^/^/g' - \\\n",
    "    | sed 's/$/,/g' \\\n",
    "    > incomplete.in.$SRA\n",
    "    \n",
    "  grep -f incomplete.in.$SRA $SRA > undone_$SRA\n",
    "  \n",
    "  rm *.tmp\n",
    "  # Still crashing due to high memory\n",
    "  # but will give a list of \"undone\" accessions\n",
    "}\n",
    "\n",
    "remove_done completed.sra inv_SraRunInfo.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make QC SRA list\n",
    "mkdir -p sra; cd sra\n",
    "mkdir -p qc\n",
    "\n",
    "aws s3 cp s3://serratus-public/tmp/completed.sra ./\n",
    "aws s3 sync s3://serratus-public/notebook/201230_sra/ ./\n",
    "\n",
    "# do not inlude prokaryotes\n",
    "cat *.csv > qc/global_SraRunInfo.csv\n",
    "\n",
    "remove_done () {\n",
    "  COMPLETE=$1\n",
    "  SRA=$2\n",
    "  \n",
    "  # Sort unique SRA in the input RunInfo file\n",
    "  cut -f1 -d',' $SRA \\\n",
    "    | sort - \\\n",
    "    > inputsra.tmp\n",
    "  \n",
    "  # Ensure completed SRA list is sorted\n",
    "  sort $COMPLETE > complete.sort.tmp\n",
    "  \n",
    "  # Supress SRA-accessions only in file 1 or file 1+2\n",
    "  # Show SRA unique to file 2\n",
    "  # parse into a regex expression to find matching lines\n",
    "  comm -1 -3 complete.sort.tmp inputsra.tmp \\\n",
    "    | sed 's/^$\\n//g' - \\\n",
    "    | sed 's/^/^/g' - \\\n",
    "    | sed 's/$/,/g' \\\n",
    "    > incomplete.in.$SRA\n",
    "    \n",
    "  grep -f incomplete.in.$SRA $SRA > undone_$SRA\n",
    "  \n",
    "  rm *.tmp\n",
    "  # Still crashing due to high memory\n",
    "  # but will give a list of \"undone\" accessions\n",
    "}\n",
    "\n",
    "remove_done completed.sra global_SraRunInfo.csv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Grouping SRA files\n",
    "\n",
    "An ordered set of queries for grouping the SRA Runs for the figure in the paper.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Human\n",
    "```\n",
    "\"Homo sapiens\"[Organism] AND (\"type_rnaseq\"[Filter] OR \"metagenomic\"[Filter] OR \"metatranscriptomic\"[Filter]) AND cluster_public[prop] AND \"platform illumina\"[Properties]\n",
    "```\n",
    "\n",
    "### Mouse\n",
    "```\n",
    "\"Mus musculus\"[orgn] AND (\"type_rnaseq\"[Filter] OR \"metagenomic\"[Filter] OR \"metatranscriptomic\"[Filter]) AND \"platform illumina\"[Properties]\n",
    "```\n",
    "\n",
    "### Mammalian\n",
    "```\n",
    "(\"Mammalia\"[Organism] NOT \"Homo sapiens\"[Organism] NOT \"Mus musculus\"[orgn]) AND (\"type_rnaseq\"[Filter] OR \"metagenomic\"[Filter] OR \"metatranscriptomic\"[Filter]) AND \"platform illumina\"[Properties]\n",
    "```\n",
    "\n",
    "### Vertebrate\n",
    "```\n",
    "(\"Vertebrata\"[Organism] NOT \"Mammalia\"[Organism] NOT \"Homo sapiens\"[Organism] NOT \"Mus musculus\"[orgn]) AND (\"type_rnaseq\"[Filter] OR \"metagenomic\"[Filter] OR \"metatranscriptomic\"[Filter]) AND \"platform illumina\"[Properties]\n",
    "```\n",
    "\n",
    "### Invertebrate\n",
    "```\n",
    "(\"Metazoa\"[Organism] NOT \"Vertebrata\"[Organism]) (\"type_rnaseq\"[Filter] OR \"metagenomic\"[Filter] OR \"metatranscriptomic\"[Filter]) AND \"platform illumina\"[Properties]\n",
    "```\n",
    "\n",
    "### Fungi\n",
    "```\n",
    "(\"Fungi\"[Organism] AND (\"type_rnaseq\"[Filter] OR \"metagenomic\"[Filter] OR \"metatranscriptomic\"[Filter]) AND \"platform illumina\"[Properties]\n",
    "```\n",
    "\n",
    "### Eukarya\n",
    "```\n",
    "(\"Eukaryota\"[Organism] NOT \"Metazoa\"[Organism] NOT \"Fungi\"[Organism]) AND (\"type_rnaseq\"[Filter] OR \"metagenomic\"[Filter] OR \"metatranscriptomic\"[Filter]) AND \"platform illumina\"[Properties]\n",
    "```\n",
    "\n",
    "### Prokaya / Other\n",
    "```\n",
    "(\"Archaea\"[Organism] OR \"Bacteria\"[Organism] AND (\"type_rnaseq\"[Filter] OR \"metagenomic\"[Filter] OR \"metatranscriptomic\"[Filter]) AND \"platform illumina\"[Properties]\n",
    "```\n",
    "\n",
    "### Metagenome\n",
    "```\n",
    "\"METAGENOMIC\" NOT amplicon[All Fields] AND \"platform illumina\"[Properties] AND cluster_public[prop]\n",
    "```\n",
    "\n",
    "### Virome\n",
    "```\n",
    "\"VIRAL METAGENOME\" OR \"VIROME\" OR \"VIROMIC\" OR \"VIRAL RNA\" OR \"VIRAL METAGENOMICS\" NOT amplicon[All Fields] AND \"platform illumina\"[Properties] AND cluster_public[prop]\n",
    "```\n",
    "\n",
    "### Mammalian WGS\n",
    "```\n",
    "(\"Mammalia\"[Organism] NOT \"Mus musculus\"[orgn]) AND (\"type_exome\"[Filter] OR \"type_genome\"[Filter]) NOT \"METAGENOME\" AND \"platform illumina\"[Properties] AND cluster_public[prop]\n",
    "```\n",
    "\n",
    "### Other\n",
    "\n",
    "- Anything in the RunInfoFiles not caught by above."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Previous Run Info Files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Human (non-controlled/dbGAP data)\n",
    "\n",
    "- File: `hu_SraRunInfo.csv`\n",
    "\n",
    "Search Term:\n",
    "```\n",
    "\"txid9606\"[Organism:exp] AND (\"type_rnaseq\"[Filter]) AND cluster_public[prop] AND \"platform illumina\"[Properties]\n",
    "```\n",
    "SRA Accessed: `2020/05/30`\n",
    "\n",
    "\n",
    "\n",
    "**AND** \n",
    "\n",
    "- File: `hu_meta_SraRunInfo.csv`\n",
    "\n",
    "Search Term: \n",
    "```\n",
    "\"txid9606\"[Organism:exp] AND (\"metagenomic\"[Filter] OR \"metatranscriptomic\"[Filter]) AND cluster_public[prop] AND \"platform illumina\"[Properties]\n",
    "```\n",
    "SRA Accessed: `2020/05/30`\n",
    "\n",
    "Results: `36103` \n",
    "\n",
    "\n",
    "### Mouse\n",
    "\n",
    "- File: `mu_SraRunInfo.csv`\n",
    "\n",
    "Search Term:\n",
    "```\n",
    "(\"Mus musculus\"[orgn]) AND (\"type_rnaseq\"[Filter] OR \"metagenomic\"[Filter] OR \"metatranscriptomic\"[Filter]) AND \"platform illumina\"[Properties]\n",
    "```\n",
    "\n",
    "SRA Accessed: `2020/06/06` \n",
    "\n",
    "Results: `890746`\n",
    "\n",
    "### Mammalian\n",
    "\n",
    "- File: `mamm_SraRunInfo.csv`\n",
    "\n",
    "Search Term:\n",
    "```\n",
    "(\"Mammalia\"[Organism] NOT \"Homo sapiens\"[Organism] NOT \"Mus musculus\"[orgn]) AND (\"type_rnaseq\"[Filter] OR \"metagenomic\"[Filter] OR \"metatranscriptomic\"[Filter]) AND \"platform illumina\"[Properties]\n",
    "```\n",
    "SRA Accessed: `2020/06/06`\n",
    "\n",
    "Results: `100798`\n",
    "\n",
    "### Vertebrate\n",
    "\n",
    "- File: `vert_SraRunInfo.csv`\n",
    "\n",
    "Search Term: \n",
    "```\n",
    "(\"Vertebrata\"[Organism] NOT \"Mammalia\"[Organism] NOT \"Homo sapiens\"[Organism] NOT \"Mus musculus\"[orgn]) AND (\"type_rnaseq\"[Filter] OR \"metagenomic\"[Filter] OR \"metatranscriptomic\"[Filter]) AND cluster_public[prop] AND \"platform illumina\"[Properties]\n",
    "```\n",
    "SRA Accessed: `2020/05/25`\n",
    "\n",
    "Results: `94908`\n",
    "\n",
    "### Vertebrate2\n",
    "\n",
    "Vertebrata and Mammalia\n",
    "Query: (\"Vertebrata\"[Organism] NOT \"Homo sapiens\"[Organism] NOT \"Mus musculus\"[orgn]) AND (\"type_rnaseq\"[Filter]) AND cluster_public[prop] AND \"platform illumina\"[Properties]\n",
    "\n",
    "Date: 200905 Return: 93981\n",
    "\n",
    "\n",
    "### Invertebrate\n",
    "\n",
    "- File: `inv_SraRunInfo.csv`\n",
    "\n",
    "Search Term: \n",
    "```\n",
    "(\"type_rnaseq\"[Filter] OR \"metagenomic\"[Filter] OR \"metatranscriptomic\"[Filter])  NOT \"Vertebrata\"[Organism] AND cluster_public[prop] AND \"platform illumina\"[Properties]\n",
    "```\n",
    "\n",
    "SRA Accessed: `2020/06/12`\n",
    "\n",
    "Results: `2193740`\n",
    "\n",
    "### Bat (update + wgs)\n",
    "\n",
    "- File: `bat_SraRunInfo.csv`\n",
    "\n",
    "Search Term: \n",
    "```\n",
    "txid9397[Organism:exp] AND \"platform illumina\"[Properties]\n",
    "```\n",
    "\n",
    "SRA Accessed: `2020/06/20`\n",
    "\n",
    "Results: `2823`\n",
    "\n",
    "### Virome\n",
    "\n",
    "- File: `viro_SraRunInfo.csv`\n",
    "\n",
    "Search Term: \n",
    "```\n",
    "\"virome\" OR \"viral metagenome\" OR \"viral metagenomics\" AND cluster_public[prop]\n",
    "```\n",
    "\n",
    "SRA Accessed: `2020/06/12`\n",
    "\n",
    "Results: `22251`\n",
    "\n",
    "### Virome2\n",
    "\n",
    "```\n",
    "Virome + Metatranscriptome\n",
    "Query: \"VIRAL METAGENOME\" OR \"VIROME\" OR \"VIROMIC\" OR \"VIRAL RNA\" OR \"METATRANSCRIPTOMIC\" NOT \"METAGENOMIC\" NOT amplicon[All Fields] AND \"platform illumina\"[Properties] AND cluster_public[prop]\n",
    "```\n",
    "\n",
    "Date: 2000904 Return: 60327\n",
    "\n",
    "### Metagenomes\n",
    "\n",
    "- File: `meta_SraRunInfo.csv`\n",
    "\n",
    "Search Term: \n",
    "```\n",
    "txid256318[Organism:noexp] #\"METATRANSCRIPTOMIC\"\n",
    "```\n",
    "\n",
    "SRA Accessed: `2020/07/12`\n",
    "\n",
    "Results: `163659`\n",
    "\n",
    "### Single-Cell RNA-seq\n",
    "Note: This is a 'negative' filter, this data-set was not analyzed it is used to separate single-cell RNA-seq from other RNA-seq libraries in downstream analysis. \n",
    "\n",
    "- File: `scRNA_SraRunInfo.csv`\n",
    "\n",
    "Search Term: \n",
    "```\n",
    "\"single-cell\" OR \"single cell\" OR \"scRNA-seq\" OR \"scRNAseq\" OR \"scRNAseq\"\n",
    "```\n",
    "\n",
    "SRA Accessed: `2020/06/22`\n",
    "\n",
    "Results: `770828`"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
